### ISSUE TITLE
Instrumentation phase failed w/ sample - no work around

### ISSUE THREAD

**Issue by nwoolls** [SteveGilham/altcover#179]
Hello there! I hope this is not just noise as this seems like a fairly common issue.

I am trying out your project as an alternative to another code coverage library but am running into an error (in fact the other library has a similar problem).

# .NET Version

```bash
dotnet --version
7.0.102
```

# Steps to Reproduce

Execute from CLI:

```bash
mkdir altcover-fails-to-resolve-mvc-razor
cd altcover-fails-to-resolve-mvc-razor

mkdir webapp
pushd webapp
dotnet new webapp
popd

mkdir xunit-tests
pushd xunit-tests
dotnet new xunit
dotnet add reference ../webapp/webapp.csproj
dotnet add package altcover
popd

dotnet new sln
dotnet sln add webapp/webapp.csproj
dotnet sln add xunit-tests/xunit-tests.csproj
```

Add the following class to the web application project:

```csharp
using Microsoft.AspNetCore.Mvc.Razor;

namespace webapp;

public static class Class1
{
    public static IMvcBuilder AddLocalization(this IMvcBuilder mvcBuilder,
        LanguageViewLocationExpanderFormat viewLocationExpanderFormat = LanguageViewLocationExpanderFormat.Suffix)
    {
        return mvcBuilder;
    }
}
```

Add the following class to the xUnit project:

```csharp
using webapp;
using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.DependencyInjection;

namespace xunit_tests;

public class Class1Tests
{
    [Fact]
    public void AddLocalizationSucceeds()
    {
        // arrange
        WebApplicationBuilder webApplicationBuilder = WebApplication.CreateBuilder();
        IMvcBuilder mvcBuilder = webApplicationBuilder.Services.AddControllersWithViews();
        
        // act
        IMvcBuilder result = mvcBuilder.AddLocalization();

        // assert
        Assert.NotNull(result);
    }
}
```

Execute from the CLI:

```bash
dotnet test /p:AltCover=true
```

# Expected Result

```
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.

Passed!  - Failed:     0, Passed:     2, Skipped:     0, Total:     2, Duration: 122 ms - xunit-tests.dll (net7.0)
  ... /Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/coverage.xml.0.acv (10,327b)
  6,188,919 visits recorded in 00:00:00.0116092 (533,104,693 visits/sec)
  A total of 6,188,919 visits recorded
  Coverage statistics flushing took 0.961 seconds
  Visited Classes 279 of 745 (37.45)
  Visited Methods 747 of 2776 (26.91)
  Visited Points 2876 of 12256 (23.47)
  Visited Branches 955 of 7956 (12)
  Maximum CRAP score 6162

  ==== Alternative Results (includes all methods including those without corresponding source) ====
  Alternative Visited Classes 291 of 816 (35.66)
  Alternative Visited Methods 839 of 3238 (25.91)
  Alternative maximum CRAP score 6162
```

# Actual Result

```
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error :  [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error : ERROR *** Instrumentation phase failed [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error :  [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error :  [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error : Failed to resolve assembly: 'Microsoft.AspNetCore.Mvc.Razor, Version=7.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error :  [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error : Details written to /Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/bin/Debug/net7.0/__Instrumented_xunit-tests/AltCover-2023-03-20--11-40-33.log [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
```

# Attempted Work Arounds

I have tried the [recommended work arounds](https://github.com/SteveGilham/altcover/wiki/QuickStart-Guide#im-now-getting-altcovertargets-error--failed-to-resolve-assembly-or-altcovertargets-warning--resolved-assembly-reference) such as:

* Adding `<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>` to the `.csproj` file(s)
* Adding `<PreserveCompilationContext>true</PreserveCompilationContext>` to the `.csproj` file(s)
* Using the `/p:CopyLocalLockFileAssemblies=true` command line argument
* Using the `/p:AltCoverDependencyList` command line argument

Unfortunately none of these resolve the issue.

In fact, trying to use the `/p:AltCoverDependencyList` argument is resulting in another error:

```bash
dotnet test /p:AltCover=true /p:CopyLocalLockFileAssemblies=true /p:AltCoverDependencyList=bin/Debug/net7.0/refs/Microsoft.AspNetCore.Mvc.Razor.dll
Microsoft (R) Test Execution Command Line Tool Version 17.4.0 (x64)
Copyright (c) Microsoft Corporation.  All rights reserved.

The argument /p:AltCover=true is invalid. Please use the /help option to check the list of valid arguments.
```

I am assuming the path is relative to the xUnit project, as I get the following if I specify a path of `foo`:

```bash
dotnet test /p:AltCover=true /p:CopyLocalLockFileAssemblies=true /p:AltCoverDependencyList=foo
  Determining projects to restore...
  All projects are up-to-date for restore.
  webapp -> /Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/webapp/bin/Debug/net7.0/webapp.dll
  xunit-tests -> /Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/bin/Debug/net7.0/xunit-tests.dll
/Users/jdoe/.nuget/packages/altcover/8.6.40/build/netstandard2.0/AltCover.targets(83,5): error : --dependency : File /Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/foo not found [/Users/jdoe/src/dotnet/altcover-fails-to-resolve-mvc-razor/xunit-tests/xunit-tests.csproj]
```

**Comment by nwoolls** [comment]
FWIW, adding the following to the xUnit `.csproj` file seems to work around the issue. Is there a better way to handle this?

```xml
<!-- 
  Begin fix / work around for error(s) generating code coverage using e.g. Coverlet or AltCover.
  
  See: https://github.com/coverlet-coverage/coverlet/issues/1459
  See: https://github.com/SteveGilham/altcover/issues/179
-->
<PropertyGroup>
  <PreserveCompilationReferences>true</PreserveCompilationReferences>
</PropertyGroup>
<Target Name="CopyReferencedAssemblies" AfterTargets="Build">
  <ItemGroup>
    <ReferencedAssemblies Include="$(OutputPath)refs/*.dll"/>
  </ItemGroup>
  
  <Copy SourceFiles="@(ReferencedAssemblies)" DestinationFolder="$(OutputPath)" ContinueOnError="false" />
</Target>
<!-- Begin fix / work around for error(s) generating code coverage using e.g. Coverlet or AltCover. -->
```

**Comment by SteveGilham** [comment]
Resolving dependencies is done in a fairly simple-minded way; by searching under the expected location of the nuget cache as 
```
Path.Combine(
      Path.Combine(
        Environment.GetFolderPath Environment.SpecialFolder.UserProfile,
        ".nuget"
      ),
      "packages"
    )
```
with a fallback of also looking in
```
Environment.GetEnvironmentVariable "NUGET_PACKAGES"
```
if that variable exists and represents a real folder path.

In this case, it seems that neither of those assumptions hold, so the program runs out of guesses as to where to find the dependencies.

I can see one obvious defect in the `NUGET_PACKAGES` fallback - it doesn't handle the case of a list of directories, if that is how it is set in your system.   Nor does the process attempt to reverse engineer how exactly it is that the program build step does the dependency resolution.

**Comment by nwoolls** [comment]
I don't have the `NUGET_PACKAGES` environment variable set.

I checked `~/.nuget` and I do see the required assembly. But the error is about 7.0.0 while the file on disk is 2.2.0:

```
ls ~/.nuget/packages/microsoft.aspnetcore.mvc.razor/2.2.0/lib/netstandard2.0
Microsoft.AspNetCore.Mvc.Razor.dll Microsoft.AspNetCore.Mvc.Razor.xml
``` 

There is no `~/.nuget/packages/microsoft.aspnetcore.mvc.razor/7.0.0` directory.

**Comment by SteveGilham** [comment]
The version 7.0 reference will be coming from the metadata in the webapp assembly.  Having dug around some more in the Razor infrastructure, it looks like the v7.0 that the program linked against is one in the SDK 7.0.x installation itself.

On Windows that's under `%PROGRAM_FILES%\dotnet\shared`, but the equivalent root used for the search on Linux is `/usr/share/dotnet/shared` (another heuristic, based on inspecting a couple of Linux installs) - so that could be where the problem lies, if that is in fact not the correct location.

**Comment by nwoolls** [comment]
@SteveGilham I'm on macOS (sorry for not mentioning earlier). The file in question looks like it's in:

```
/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App/7.0.2
```

```bash
ls /usr/local/share/dotnet/shared/Microsoft.AspNetCore.App/7.0.2/Microsoft.AspNetCore.Mvc.Razor.dll
/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App/7.0.2/Microsoft.AspNetCore.Mvc.Razor.dll
```

**Comment by SteveGilham** [comment]
So, easy enough to add `/usr/local/share` to my list of places to try.  Before I do that and put out another release, I'll review what the actual MSBuild process does to find the location and see if there's something I can draw on from there that is better than just a list of "it has been observed in this location".

**Comment by SteveGilham** [comment]
Here's a pre-release build with the results of my delving into the way MSBuild resolves things.  If you could give it a spin, then I'll make a full release if it's good.

[altcover.8.6.44-pre-gac0509837d.nupkg.zip](https://github.com/SteveGilham/altcover/files/11032474/altcover.8.6.44-pre-gac0509837d.nupkg.zip)

**Comment by SteveGilham** [comment]
Super. Look out for a new release soon.

**Comment by SteveGilham** [comment]
Should be fixed in [release 8.6.45](https://www.nuget.org/packages/altcover/8.6.45#release-body-tab).
